src/Simplex/Messaging/Agent.hs

```diff
@@ -1056,22 +1083,32 @@ enqueueMessages c cData sqs msgFlags aMessage = do
   enqueueMessages' c cData sqs msgFlags aMessage
 
 enqueueMessages' :: AgentMonad m => AgentClient -> ConnData -> NonEmpty SndQueue -> MsgFlags -> AMessage -> m AgentMsgId
-enqueueMessages' c cData (sq :| sqs) msgFlags aMessage = do
-  msgId <- enqueueMessage c cData sq msgFlags aMessage
-  mapM_ (enqueueSavedMessage c cData msgId) $
-    filter (\SndQueue {status} -> status == Secured || status == Active) sqs
-  pure msgId
+enqueueMessages' c cData sqs msgFlags aMessage =
+  oneResult $ \r -> enqueueMessagesB c [(r, (cData, sqs, msgFlags, aMessage))]
+
+enqueueMessagesB :: AgentMonad' m => AgentClient -> [(EIORef AgentMsgId, (ConnData, NonEmpty SndQueue, MsgFlags, AMessage))] -> m ()
+enqueueMessagesB _ [] = pure ()
+enqueueMessagesB c reqs = enqueueMessageB c reqs >>= enqueueSavedMessageB c
+
+isActiveSndQ :: SndQueue -> Bool
+isActiveSndQ SndQueue {status} = status == Secured || status == Active
 
 enqueueMessage :: forall m. AgentMonad m => AgentClient -> ConnData -> SndQueue -> MsgFlags -> AMessage -> m AgentMsgId
-enqueueMessage c cData@ConnData {connId} sq msgFlags aMessage = do
-  resumeMsgDelivery c cData sq
+enqueueMessage c cData sq msgFlags aMessage =
+  oneResult $ \r -> enqueueMessageB c [(r, (cData, [sq], msgFlags, aMessage))]
+
+-- this function is used only for sending messages in batch, it returns the list of successes to enqueue additional deliveries
+enqueueMessageB :: forall m. AgentMonad' m => AgentClient -> [(EIORef AgentMsgId, (ConnData, NonEmpty SndQueue, MsgFlags, AMessage))] -> m [(ConnData, [SndQueue], AgentMsgId)]
+enqueueMessageB c reqs = do
+  forM_ reqs $ \(_, (cData, sq :| _, _, _)) ->
+    runExceptT $ resumeMsgDelivery c cData sq
```

Comment:
wouldn't this call `resumeMsgDelivery` multiple time for same send queue?

It may not be a problem, but seems excessive in scope of one function, we could accumulate "already resumed" in scope of function

Reply:
it seems a simple check is inside the function so it's probably ok